[WGSL] Add support for binary add

tadeuzagallo opened the pull request:
#### 243c46b4a2b383e9224f4794a489bde1cd185a9a
<pre>
[WGSL] Add support for binary add
<a href="https://bugs.webkit.org/show_bug.cgi?id=250849">https://bugs.webkit.org/show_bug.cgi?id=250849</a>
&lt;rdar://problem/104432245&gt;

Reviewed by Myles C. Maxfield.

Add lexing for `+` and `++`, add a new BinaryExpression AST node,
implement parsing and codegen for add.

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTBinaryExpression.h: Copied from Source/WebGPU/WGSL/AST/AST.h.
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer&lt;T&gt;::lex):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseArrayType):
(WGSL::Parser&lt;Lexer&gt;::parseReturnStatement):
(WGSL::Parser&lt;Lexer&gt;::parseRelationalExpression):
(WGSL::Parser&lt;Lexer&gt;::parseShiftExpression):
(WGSL::Parser&lt;Lexer&gt;::parseAdditiveExpression):
(WGSL::Parser&lt;Lexer&gt;::parseMultiplicativeExpression):
(WGSL::Parser&lt;Lexer&gt;::parseExpression):
(WGSL::Parser&lt;Lexer&gt;::parseShortCircuitOrExpression): Deleted.
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Token.cpp:
(WGSL::toString):
* Source/WebGPU/WGSL/Token.h:
* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp:
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/259133@main">https://commits.webkit.org/259133@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a6c33364d6fbab302b84d5aecffeeb27c1d9c9b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13123 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113239 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173545 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107972 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14180 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4035 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96261 "Reverted pull request changes (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112323 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109798 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93992 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96261 "Reverted pull request changes (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25604 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96261 "Reverted pull request changes (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6492 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26981 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6637 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3518 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12634 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46515 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6293 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8412 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/AST/ASTBinaryExpression.h:
```diff
@@ -0,0 +1,60 @@
+namespace WGSL::AST {
+
+enum class BinaryOperation : uint8_t {
+    Add,
```
heh, maybe add a FIXME for the others? :P

tadeuzagallo replied:
I can, but I don't think it's necessary. We already have FIXMEs in the parser, which means we'll inevitably have to fill these up.

litherum commented on Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
```diff
@@ -265,6 +265,15 @@ void StringDumper::visit(UnaryExpression& expression)
 
+void StringDumper::visit(BinaryExpression& expression)
+{
+    constexpr ASCIILiteral binaryOperator[] = { "+"_s };
```
Is there a way to enforce that these stay in the right order?

tadeuzagallo replied:
There is, but I think we should eventually just use a switch case, but I mostly just followed what `UnaryExpression` is doing.

litherum commented on Source/WebGPU/WGSL/Lexer.cpp:
```diff
@@ -108,6 +108,14 @@ Token Lexer<T>::lex()
         }
         return makeToken(TokenType::Minus);
         break;
+    case '+':
```
I assume this won't choke on `++`?

djg replied:
'++' is checked two lines below.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -316,6 +317,16 @@ void FunctionDefinitionWriter::visit(AST::UnaryExpression& unary)
 
+void FunctionDefinitionWriter::visit(AST::BinaryExpression& binary)
+{
+    visit(binary.lhs());
```
No parentheses? I don't think our binary operator precedence is exactly the same as C's, IIRC.

djg replied:
I think that we should eventually write these out in "assembly-as-C" form. Eg:

```t0 = <lhs>;
t1 = <rhs>;```
```t2 = t0 + t1;```

Does that make sense?

tadeuzagallo replied:
Yes, I agree, at first I wasn't sure, but given the differences in precedence and semantics, I think that's probably easier. I think we should just implement that in a separate pass once we have more operators.

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Add support for binary add
https://bugs.webkit.org/show_bug.cgi?id=250849
<rdar://problem/104432245>

Reviewed by Myles C. Maxfield.

Add lexing for `+` and `++`, add a new BinaryExpression AST node,
implement parsing and codegen for add.

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTBinaryExpression.h: Copied from Source/WebGPU/WGSL/AST/AST.h.
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer<T>::lex):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseArrayType):
(WGSL::Parser<Lexer>::parseReturnStatement):
(WGSL::Parser<Lexer>::parseRelationalExpression):
(WGSL::Parser<Lexer>::parseShiftExpression):
(WGSL::Parser<Lexer>::parseAdditiveExpression):
(WGSL::Parser<Lexer>::parseMultiplicativeExpression):
(WGSL::Parser<Lexer>::parseExpression):
(WGSL::Parser<Lexer>::parseShortCircuitOrExpression): Deleted.
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Token.cpp:
(WGSL::toString):
* Source/WebGPU/WGSL/Token.h:
* Tools/TestWebKitAPI/Tests/WGSL/LexerTests.cpp:
(TestWGSLAPI::TEST):
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):

Canonical link: https://commits.webkit.org/259133@main
